### DISCUSSION TITLE
Bananas - Python client?

### DISCUSSION THREAD

**Discussion by michalc** [OpenTTD/OpenTTD#11863] [Q&A] [answered]
I'm writing something I'm calling OpenTTDLab https://github.com/michalc/OpenTTDLab (_really_ early version: proof of concept really) to run sets of experiments using OpenTTD from Python, e.g. from JupyterLab.

I would like to make it really easy to use it with AIs published at https://bananas.openttd.org/, so without having to even fire up OpenTTD "manually" to download them.

Would it be acceptable for me to write a Python client to do this? I note at https://bananas.openttd.org/ it does say "For downloading content, please use OpenTTD's built-in content downloader", so I'm not sure if I should really.

**Comment by TrueBrain** [comment, marked as answer]
What a cool project! Do keep us updated on how it is going, and how you are using it.

For your questions, there are a few possible answer. We have been willing to add in-game the ability to automatically download all missing content from BaNaNaS, if you load a game that is missing any. So you can just do `./openttd -g <savegame> -DOWNLOADSTUFF`, and that it makes it happen. Via the UI it is already possible, but not in-game yet.
But, as that goes, nobody actually implemented that yet, so we don't have that yet.

For downloading yourself via a Python client or what-ever, there isn't any issue to do so. Our ToS allows you to download the latest version of content; and although the idea of course is that it is done via the OpenTTD client, that can't be guarded.

That said, if you go that path, I do have three requests:
1) cache as much as possible. Please do not download on every run all the files you need. Downloads costs us money, we live on donations, so please keep it reasonable.
2) please use the HTTPS to download the actual files. Although the initial handshake is via a custom TCP protocol, once you got the IDs of the files you are looking for, there is a dance via HTTPS to get the actual download URLs. Please invest time in using this, over downloading via TCP (although that is also possible). The reason here again is money. Downloads via TCP are handled via AWS, for which we pay 0,09 dollar per GB transferred. It is REALLY expensive.
3) please never download ALL content from BaNaNaS. Although there are a few hurdles to prevent you to do that accidentally, please also don't try to look for ways to do this. Again the reason here is cost.

In other words: be a good neighbour, and go nuts. Be a bad neighbour .. and I might get mad :P

If you need any help deciphering how the TCP protocol and HTTPS dance works, let me know. Easiest is via Discord, but via this Discussion is also fine.

Happy hacking!

**Reply by michalc** [reply]
This is incredibly helpful.

The requests are very reasonable: will of course abide (and have implemented _some_ caching already in OpenTTDLab, but more to be done).

> Do keep us updated on how it is going, and how you are using it.

Will do!

> If you need any help deciphering how the TCP protocol and HTTPS dance works, let me know

I may take you up on that :-)

Thank you!!

**Reply by jasfvcb** [reply]
666

**Reply by michalc** [reply]
Have now implemented a basic/rough version: https://github.com/michalc/OpenTTDLab/blob/60f848a30073b92e012837177d642f5030f4ff86/openttdlab.py#L272

**Reply by TrueBrain** [reply]
Looking good :)

**Comment by TrueBrain** [comment]
To maybe inspire you a bit what we have been wanting for years now, but never really got the time to build, or didn't take the time to build:

When we release a new OpenTTD version, it is very hard for us to see if any regression creeped in. Yes, we have unit tests. Yes, we have regression tests. But still .. regressions can happen, and over 3 axis:

- CPU / memory degradation (lower game fps.
- Rendering degradation (lower video fps)
- Crashing savegames that used to work.

For the first, I wrote some scripts last year, where I run a whole set of savegames from time to time to benchmark their performance. It is just a collection of scripts linked together, outputting a csv with CPU / memory usage for each game. This I load into Grafana to create graphs. Source can be found here: https://github.com/TrueBrain/OpenTTD-performance

For the second, we have nothing. Well, yes, our own experience when starting the game. But nothing systematic, or which tells us what commit made it better / worse.

The last one, once upon a time I started https://github.com/TrueBrain/OpenTTD-savegames, and LordAro also gave it a few gos, but never anything came from it. From all three, this for us is the most important one: to know if a new commit / version breaks savegames that used to work.

For all this, we kinda miss a framework to help us do this. We can manage the compute unit (VM or whatever) to run it on, but we do not have any tooling assisting us running this every commit, or every week, or whatever.

Then I do not know your actual goal of OpenTTDLab, but just in case you are looking for some inspiration what us, OpenTTD Developers, would really help out .. an idea / insight in what is troubling us. Anything you can do in that regard to help us out, would be greatly appreciated. Even if it is: this tool isn't actually designed for it, but if you do A B C, you can use it for that goal, would be perfect.

Not saying you need to do anything with this comment; but if it so happens to align with anything you are doing, do let me know please :)

**Reply by michalc** [reply]
So my envisioned use is a bit different... it's more on the science-y/research side of things, especially AIs. I'm wanting to make it as easy as possible to replicate/reproduce research results that use OpenTTD. There are a few published papers that use OpenTTD, but it's hard to reproduce their results (or even in some cases I suspect, for the authors to replicate their own results because it's clear there were manual steps involved)

Or at least, do that for a few simple cases/take steps towards it being easier.

But: I think there is (or rather, could be) overlap between our goals: both of these things are all about reproducibility/determinism... I will have a ponder. I did want to make OpenTTDLab have lots of defaults but then also be flexible, so maybe that flexibility could lend itself to regression tests...

**Reply by michalc** [reply]
> Do keep us updated on how it is going, and how you are using it.

So not exactly the regressions you were talking about, but have an example of how it can be used for regression tests of AIs. e.g https://github.com/michalc/ParameterisedAI/blob/main/test_parameterised_ai.py, and run in a GitHub actions workflow: https://github.com/michalc/ParameterisedAI/blob/main/.github/workflows/test.yml

**Reply by michalc** [reply]
So I now have a more "public" BaNaNaS Python client exposed as part of https://github.com/michalc/OpenTTDLab. For example:

```python
from openttdlab import download_from_bananas

with download_from_bananas('ai/41444d4c') as files:
    for content_id, filename, md5_partial, get_data in files:
        with get_data() as chunks:
            with open(filename, 'wb') as f:
                for chunk in chunks:
                    f.write(chunk)
```
It fetches all direct and transitive dependencies for the latest version of something published to BaNaNaS.

I would like to make it be able to fetch the non-latest version: so it would return the _full_ MD5 of something published, and then accept that full MD5 as an argument to re-fetch the exact same thing. But would this essentially fall into the definition of being a bad neighbour, given the BaNaNaS TOS?

**Reply by michalc** [reply]
> I would like to make it be able to fetch the non-latest version: so it would return the full MD5 of something published, and then accept that full MD5 as an argument to re-fetch the exact same thing.

I'm tempted to allow this behaviour... but only for content that has a suitable license (which right now I suspect is everything but "Custom"). Because then the license given by the author grants its distribution (subject to maybe other conditions specific to the license), so it's not helping anyone do anything that the original author didn't give permission for.

Shout if I shouldn't!

**Reply by michalc** [reply]
@TrueBrain 

> Do keep us updated on how it is going, and how you are using it.

Just in this vein / if you or others are interested:

- There have been quite a few versions/updates to https://github.com/michalc/OpenTTDLab (although there is always loads more to do)
- This whole thing was for an MSc dissertation project, which is available (hopefully "forever") at https://doi.org/10.7488/era/5414
- And I've got a WIP short paper for [JOSS](https://joss.theoj.org/) at https://joss.theoj.org/papers/993f4e4e559f9ca7ead12d4511430dba (not much there right now, but it's in "pre-review" over at https://github.com/openjournals/joss-reviews/issues/7600, and you can see the draft at https://github.com/openjournals/joss-papers/blob/joss.07600/joss.07600/10.21105.joss.07600.pdf)

**Reply by michalc** [reply]
> Do keep us updated on how it is going, and how you are using it.

And the paper has just today been published! https://joss.theoj.org/papers/10.21105/joss.08014
